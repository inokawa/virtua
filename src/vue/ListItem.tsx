/** @jsxImportSource vue */
import {
  ref,
  defineComponent,
  watch,
  StyleValue,
  PropType,
  VNode,
  NativeElements,
} from "vue";
import { ItemResizeObserver } from "../core/resizer";
import { isRTLDocument } from "../core/environment";

/**
 * @internal
 */
export const ListItem = /*#__PURE__*/ defineComponent({
  props: {
    _children: { type: Object as PropType<VNode>, required: true },
    _resizer: {
      type: Function as PropType<ItemResizeObserver>,
      required: true,
    },
    _index: { type: Number, required: true },
    _offset: { type: Number, required: true },
    _hide: { type: Boolean },
    _isHorizontal: { type: Boolean },
    _isSSR: { type: Boolean },
    _as: { type: String as PropType<keyof NativeElements>, required: true },
  },
  setup(props) {
    const elementRef = ref<HTMLDivElement>();

    // The index may be changed if elements are inserted to or removed from the start of props.children
    watch(
      () => elementRef.value && props._index,
      (_, __, onCleanup) => {
        onCleanup(props._resizer(elementRef.value!, props._index));
      },
      {
        flush: "post",
      }
    );

    return () => {
      const {
        _children: children,
        _offset: offset,
        _hide: hide,
        _isHorizontal: isHorizontal,
        _isSSR: isSSR,
        _as: Element,
      } = props;

      const style: StyleValue = {
        position: hide && isSSR ? undefined : "absolute",
        [isHorizontal ? "height" : "width"]: "100%",
        [isHorizontal ? "top" : "left"]: "0px",
        [isHorizontal ? (isRTLDocument() ? "right" : "left") : "top"]:
          offset + "px",
        visibility: !hide || isSSR ? "visible" : "hidden",
      };
      if (isHorizontal) {
        style.display = "flex";
      }

      return (
        <Element ref={elementRef} style={style}>
          {children}
        </Element>
      );
    };
  },
});
