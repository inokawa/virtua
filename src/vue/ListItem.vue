<script setup lang="ts">
import {
    ref,
    watch,
    NativeElements,
    computed,
    Ref,
} from "vue";
import {
    ItemResizeObserver,
    isRTLDocument,
    StateVersion,
    VirtualStore,
} from "../core/index.js";
import type { ItemProps } from "./utils.js";

const props = defineProps<{
    _stateVersion: number,
    _store: VirtualStore,
    _children: VNode,
    _resizer: ItemResizeObserver,
    _index: number,
    _isHorizontal: boolean,
    _isSSR: boolean;
    _as: keyof NativeElements,
    _itemProps: ReturnType<ItemProps>
}>();

const elementRef = ref<HTMLDivElement>();

const offset = computed(() =>
    props._stateVersion && props._store.$getItemOffset(props._index)
);
const hide = computed(() =>
    props._stateVersion && props._store.$isUnmeasuredItem(props._index)
);

// The index may be changed if elements are inserted to or removed from the start of props.children
watch(
    () => elementRef.value && props._index,
    (_, __, onCleanup) => {
        onCleanup(props._resizer(elementRef.value!, props._index));
    },
    {
        flush: "post",
    }
);
</script>

<template>
    <component :is="props._as" ref="elementRef" :style="{
        contain: 'layout style',
        position: hide && props._isSSR ? undefined : 'absolute',
        [props._isHorizontal ? 'height' : 'width']: '100%',
        [props._isHorizontal ? 'top' : 'left']: '0px',
        [props._isHorizontal ? (isRTLDocument() ? 'right' : 'left') : 'top']: offset + 'px',
        visibility: !hide || props._isSSR ? 'visible' : 'hidden',
        ...(props._itemProps?.style ?? {}),
        display: props._isHorizontal ? 'inline-flex' : undefined,
    }" v-bind="props._itemProps">
        <component :is="props._children" />
    </component>
</template>
