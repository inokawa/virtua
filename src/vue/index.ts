/**
 * @module vue
 */
export { default as VList } from "./VList.vue";
export type { VListHandle } from "./VList.vue";
export { default as Virtualizer } from "./Virtualizer.vue";
export type { VirtualizerHandle } from "./Virtualizer.vue";
export { default as WindowVirtualizer } from "./WindowVirtualizer.vue";
export type { WindowVirtualizerHandle } from "./WindowVirtualizer.vue";
