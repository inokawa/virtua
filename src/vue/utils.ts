import { CSSProperties, VNode } from "vue";
import { ItemsRange } from "../core";

/**
 * @internal
 */
export const getKey = (e: VNode, i: number): Exclude<VNode["key"], null> => {
  const key = e.key;
  return key != null ? key : "_" + i;
};

/**
 * @internal
 */
export const isSameRange = (prev: ItemsRange, next: ItemsRange): boolean => {
  return prev[0] === next[0] && prev[1] === next[1];
};

export type ItemProps = (payload: {
  item: any;
  index: number;
}) => { [key: string]: any; style?: CSSProperties; class?: string } | undefined;
