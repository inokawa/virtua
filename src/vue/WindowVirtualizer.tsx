/** @jsxImportSource vue */
import {
  ref,
  onMounted,
  defineComponent,
  onUnmounted,
  VNode,
  watch,
  ComponentOptionsMixin,
  SlotsType,
  ComponentOptionsWithObjectProps,
  ComponentObjectPropsOptions,
  PropType,
  NativeElements,
  computed,
} from "vue";
import {
  UPDATE_SCROLL_END_EVENT,
  UPDATE_VIRTUAL_STATE,
  createVirtualStore,
  ACTION_ITEMS_LENGTH_CHANGE,
  UPDATE_SCROLL_EVENT,
} from "../core/store";
import { createWindowResizer } from "../core/resizer";
import { createWindowScroller } from "../core/scroller";
import { ListItem } from "./ListItem";
import { getKey, isSameRange } from "./utils";
import { ItemsRange, ScrollToIndexOpts } from "../core/types";

export interface WindowVirtualizerHandle {
  /**
   * Find the start index of visible range of items.
   */
  findStartIndex: () => number;
  /**
   * Find the end index of visible range of items.
   */
  findEndIndex: () => number;
  /**
   * Scroll to the item specified by index.
   * @param index index of item
   * @param opts options
   */
  scrollToIndex(index: number, opts?: ScrollToIndexOpts): void;
}

const props = {
  /**
   * The data items rendered by this component.
   */
  data: { type: Array, required: true },
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan: Number,
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize: Number,
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift: Boolean,
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal: Boolean,
  /**
   * Component or element type for container element.
   * @defaultValue "div"
   */
  as: { type: String as PropType<keyof NativeElements>, default: "div" },
  /**
   * Component or element type for item element.
   * @defaultValue "div"
   */
  item: { type: String as PropType<keyof NativeElements>, default: "div" },
} satisfies ComponentObjectPropsOptions;

export const WindowVirtualizer = /*#__PURE__*/ defineComponent({
  props,
  emits: ["scroll", "scrollEnd"],
  setup(props, { emit, slots, expose }) {
    const isHorizontal = props.horizontal;
    const containerRef = ref<HTMLDivElement>();
    const store = createVirtualStore(
      props.data.length,
      props.itemSize,
      props.overscan,
      undefined,
      undefined,
      !props.itemSize
    );
    const resizer = createWindowResizer(store, isHorizontal);
    const scroller = createWindowScroller(store, isHorizontal);

    const rerender = ref(store._getStateVersion());
    const unsubscribeStore = store._subscribe(UPDATE_VIRTUAL_STATE, () => {
      rerender.value = store._getStateVersion();
    });

    const unsubscribeOnScroll = store._subscribe(UPDATE_SCROLL_EVENT, () => {
      emit("scroll", store._getScrollOffset());
    });
    const unsubscribeOnScrollEnd = store._subscribe(
      UPDATE_SCROLL_END_EVENT,
      () => {
        emit("scrollEnd");
      }
    );

    const range = computed<ItemsRange>((prev) => {
      rerender.value;
      const next = store._getRange();
      if (prev && isSameRange(prev, next)) {
        return prev;
      }
      return next;
    });
    const isScrolling = computed(() => rerender.value && store._isScrolling());
    const totalSize = computed(() => rerender.value && store._getTotalSize());
    const jumpCount = computed(() => rerender.value && store._getJumpCount());

    onMounted(() => {
      const el = containerRef.value;
      if (!el) return;
      resizer._observeRoot(el);
      scroller._observe(el);
    });
    onUnmounted(() => {
      unsubscribeStore();
      unsubscribeOnScroll();
      unsubscribeOnScrollEnd();
      resizer._dispose();
      scroller._dispose();
    });

    watch(
      () => props.data.length,
      (count) => {
        store._update(ACTION_ITEMS_LENGTH_CHANGE, [count, props.shift]);
      }
    );

    watch(
      [jumpCount],
      () => {
        scroller._fixScrollJump();
      },
      { flush: "post" }
    );

    expose({
      findStartIndex: store._findStartIndex,
      findEndIndex: store._findEndIndex,
      scrollToIndex: scroller._scrollToIndex,
    } satisfies WindowVirtualizerHandle);

    return () => {
      const Element = props.as;
      const ItemElement = props.item;

      const [startIndex, endIndex] = range.value;
      const total = totalSize.value;

      const items: VNode[] = [];
      for (let i = startIndex, j = endIndex; i <= j; i++) {
        const e = slots.default({ item: props.data![i]!, index: i })[0]!;
        items.push(
          <ListItem
            key={getKey(e, i)}
            _rerender={rerender}
            _store={store}
            _resizer={resizer._observeItem}
            _index={i}
            _children={e}
            _isHorizontal={isHorizontal}
            _as={ItemElement}
          />
        );
      }

      return (
        <Element
          ref={containerRef}
          style={{
            // contain: "content",
            overflowAnchor: "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
            flex: "none", // flex style can break layout
            position: "relative",
            visibility: "hidden", // TODO replace with other optimization methods
            width: isHorizontal ? total + "px" : "100%",
            height: isHorizontal ? "100%" : total + "px",
            pointerEvents: isScrolling.value ? "none" : undefined,
          }}
        >
          {items}
        </Element>
      );
    };
  },
} as ComponentOptionsWithObjectProps<
  typeof props,
  void,
  {},
  {},
  {},
  ComponentOptionsMixin,
  ComponentOptionsMixin,
  {
    /**
     * Callback invoked whenever scroll offset changes.
     * @param offset Current scrollTop, or scrollLeft if horizontal: true.
     */
    scroll: (offset: number) => void;
    /**
     * Callback invoked when scrolling stops.
     */
    scrollEnd: () => void;
  },
  string,
  {},
  string,
  SlotsType<{ default: (arg: { item: any; index: number }) => VNode[] }>
>);
