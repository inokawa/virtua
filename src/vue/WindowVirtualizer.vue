<script setup lang="ts" generic="T">
import {
  ref,
  onMounted,
  onUnmounted,
  watch,
  NativeElements,
  computed,
  defineEmits,
  defineProps,
  useSlots,
  defineExpose,
  VNode,
} from "vue";
import {
  UPDATE_SCROLL_END_EVENT,
  UPDATE_VIRTUAL_STATE,
  createVirtualStore,
  ACTION_ITEMS_LENGTH_CHANGE,
  UPDATE_SCROLL_EVENT,
  createWindowResizer,
  createWindowScroller,
  ItemsRange,
  ScrollToIndexOpts,
} from "../core/index.js";
import { default as ListItem } from "./ListItem.vue";
import { getKey, isSameRange } from "./utils.js";

export interface WindowVirtualizerHandle {
  /**
   * Find the start index of visible range of items.
   */
  findStartIndex: () => number;
  /**
   * Find the end index of visible range of items.
   */
  findEndIndex: () => number;
  /**
   * Scroll to the item specified by index.
   * @param index index of item
   * @param opts options
   */
  scrollToIndex(index: number, opts?: ScrollToIndexOpts): void;
}

interface WindowVirtualizerProps<T> {
  /**
   * The data items rendered by this component.
   */
  data: T[];
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan?: number;
  /**
   * Item size hint for unmeasured items in pixels. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize?: number;
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift?: boolean;
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal?: boolean;
  /**
   * Component or element type for container element.
   * @defaultValue "div"
   */
  as?: keyof NativeElements;
  /**
   * Component or element type for item element.
   * @defaultValue "div"
   */
  item?: keyof NativeElements;
}

defineSlots<{
  default: (props: { item: T, index: number }) => VNode;
}>();

const props = withDefaults(defineProps<WindowVirtualizerProps<T>>(), { overscan: 4, as: 'div', item: 'div' });
const emit = defineEmits<{ scroll: []; scrollEnd: [] }>();
const slots = useSlots();

const isHorizontal = !!props.horizontal;
const containerRef = ref<HTMLDivElement>();
const store = createVirtualStore(
  props.data.length,
  props.itemSize,
  props.overscan,
  undefined,
  undefined,
  !props.itemSize
);
const resizer = createWindowResizer(store, isHorizontal);
const scroller = createWindowScroller(store, isHorizontal);

const stateVersion = ref(store.$getStateVersion());
const unsubscribeStore = store.$subscribe(UPDATE_VIRTUAL_STATE, () => {
  stateVersion.value = store.$getStateVersion();
});
const unsubscribeOnScroll = store.$subscribe(UPDATE_SCROLL_EVENT, () => {
  // https://github.com/inokawa/virtua/discussions/580
  emit("scroll");
});
const unsubscribeOnScrollEnd = store.$subscribe(
  UPDATE_SCROLL_END_EVENT,
  () => {
    emit("scrollEnd");
  }
);

const range = computed<ItemsRange>((prev) => {
  stateVersion.value;
  const next = store.$getRange();
  if (prev && isSameRange(prev, next)) {
    return prev;
  }
  return next;
});
const isScrolling = computed(() => stateVersion.value && store.$isScrolling());
const totalSize = computed(() => stateVersion.value && store.$getTotalSize());

onMounted(() => {
  const el = containerRef.value;
  if (!el) return;
  resizer.$observeRoot(el);
  scroller.$observe(el);
});
onUnmounted(() => {
  unsubscribeStore();
  unsubscribeOnScroll();
  unsubscribeOnScrollEnd();
  resizer.$dispose();
  scroller.$dispose();
});

watch(
  () => props.data.length,
  (count) => {
    store.$update(ACTION_ITEMS_LENGTH_CHANGE, [count, props.shift]);
  }
);
watch(
  [stateVersion],
  () => {
    scroller.$fixScrollJump();
  },
  { flush: "post" }
);

defineExpose<WindowVirtualizerHandle>({
  findStartIndex: store.$findStartIndex,
  findEndIndex: store.$findEndIndex,
  scrollToIndex: scroller.$scrollToIndex,
});

const style = computed(() => {
  const total = totalSize.value
  return {
    contain: 'size paint style',// https://github.com/inokawa/virtua/pull/775
    overflowAnchor: "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
    overflow: "clip", // https://github.com/inokawa/virtua/pull/485 https://github.com/inokawa/virtua/issues/717
    flex: "none", // flex style can break layout
    position: 'relative',
    width: isHorizontal ? total + 'px' : '100%',
    height: isHorizontal ? '100%' : total + 'px',
    pointerEvents: isScrolling.value ? 'none' : undefined,
  }
});


const getSlot = (index: number) => {
  return slots["default"]?.({ item: props.data?.[index], index })?.[0] ?? null;
}
const getKeyForItem = (index: number) => {
  const slot = getSlot(index);
  return slot ? getKey(slot, index) : index;
}

const rangeArray = computed(() => {
  const indexes: number[] = [];
  for (let [i, j] = range.value; i <= j; i++) {
    indexes.push(i);
  }

  return indexes
});
</script>

<template>
  <component :is="props.as" ref="containerRef" :style="style">
    <template v-for="index in rangeArray" :is="props.item" :key="getKeyForItem(index)" v-bind="getItemProps(index)">
      <ListItem :_stateVersion="stateVersion" :_store="store" :_resizer="resizer.$observeItem" :_index="index"
        :_children="getSlot(index)" :_isHorizontal="isHorizontal" :_isSSR="isSSR" :_as="props.item"
        :_itemProps="props.itemProps?.({ item: props.data[index], index })" />
    </template>
  </component>
</template>
