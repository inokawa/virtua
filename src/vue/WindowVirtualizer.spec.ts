import { it, expect, describe } from "vitest";
import { defineComponent, h } from "vue";
import { render } from "@testing-library/vue";
import { WindowVirtualizer } from "./WindowVirtualizer";
import { setupJsDomEnv } from "../../scripts/spec";

setupJsDomEnv({
  itemWidth: 100,
  itemHeight: 50,
});

const range = (length: number) => Array.from({ length }).map((_, i) => i);

it("should pass attributes to element", () => {
  const wrapper = render(WindowVirtualizer, {
    props: {
      data: range(1),
    },
    attrs: {
      id: "id",
      className: "class",
      "tab-index": 0,
      role: "list",
      "aria-label": "test",
      style: { background: "red" },
    },
    slots: { default: (data: any) => h("div", { key: data }, data) },
  });
  expect(wrapper.baseElement).toMatchSnapshot();
});

describe("vertical", () => {
  it("should render 0 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: [],
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(1),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(5),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(100),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(1000),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(10000),
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = defineComponent({
      props: ["data"],
      setup(props) {
        return () => {
          return h("div", props.data);
        };
      },
    });
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(3),
      },
      slots: { default: (data: any) => h(Comp, { key: data, data }) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});

describe("horizontal", () => {
  it("should render 0 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: [],
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(1),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(5),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(100),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(1000),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(10000),
        horizontal: true,
      },
      slots: { default: (data: any) => h("div", { key: data }, data) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = defineComponent({
      props: ["data"],
      setup(props) {
        return () => {
          return h("div", props.data);
        };
      },
    });
    const wrapper = render(WindowVirtualizer, {
      props: {
        data: range(3),
        horizontal: true,
      },
      slots: { default: (data: any) => h(Comp, { key: data, data }) },
    });
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
