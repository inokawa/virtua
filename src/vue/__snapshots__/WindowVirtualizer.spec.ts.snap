// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`horizontal > should render 0 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 0px; height: 100%;"
    />
  </div>
</body>
`;

exports[`horizontal > should render 1 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 40px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
      >
        <div>
          0
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 5 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 200px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 100 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 4000px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 1000 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 40000px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render 10000 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 400000px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 120px; visibility: hidden; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 160px; visibility: hidden; display: flex;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`horizontal > should render component 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 120px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: hidden; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 40px; visibility: hidden; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 80px; visibility: hidden; display: flex;"
      >
        <div>
          2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should pass attributes to element 1`] = `
<body>
  <div>
    <div
      aria-label="test"
      class="class"
      id="id"
      role="list"
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 40px; background: red;"
      tab-index="0"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 0 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 0px;"
    />
  </div>
</body>
`;

exports[`vertical > should render 1 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 40px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 5 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 200px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 100 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 4000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 1000 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 40000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render 10000 children 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 400000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`vertical > should render component 1`] = `
<body>
  <div>
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 120px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
      >
        <div>
          2
        </div>
      </div>
    </div>
  </div>
</body>
`;
